README.md
Graphite
========

[![Build Status](https://travis-ci.org/issuu/graphite-async.svg?branch=master)](https://travis-ci.org/issuu/graphite-async)

Report performance stats (with percentiles) to Graphite over AMQP.

Dependencies
------------

  * OPAM 2
  * `dune-release` (for releases)

Usage
-----

It is a normal `dune` project, can be built in all normal ways `dune`
software can be built. The `make help` command provides a list of possibly
useful actions.

Release process
---------------

Releasing a new version is handled by `dune-release`, check the
[Mirage instructions][mirage] on their packaging process.

1. Write changelog entry
1. `make tag` to tag the repository locally
1. `make distrib` to create a release tarball
1. `make publish` to push the tag and create a GitHub release, upload the tarball
   and set the changelog.

The publish step requires access to a [personal API token][ghapi] on GitHub
that has the `repo` scope.  Possibly only `public_repo` but it is a bit
unclear.

[mirage]: https://mirage.io/wiki/packaging
[ghapi]: https://github.com/settings/tokens


CHANGES.md
1.0.0
=====

* Remove optional Graphite.t parameters

0.14.3
======

* Add support for Mock graphite that does not report to remote

0.14.2
======

* Add support for callback functionality on timing functions.

0.14.1
======

* Do not lock dependencies to versions in advance. In particular the
  `ppx_deriving` lockdown prevents use of OCaml 4.08.

0.14.0
======

* Expose `flush` endpoint to flush queued message to graphite

0.13.2
======

* Add `Deferred.keyed_time` to allow deferred to decide on part of the key
  string. Useful for logging per status code or similar.

0.13.1
======

* Drop upper version constraints on Jane Street packages

0.13.0
======

* Add `sum` to report
* Add `obs_rate` to report. Shows the rate of observations.
* `rate` now shows the rate of change in percentile sums.

0.12.0
======

* Fix percentile reports reporting sums as observations.
* Add `avg` to percentiles in reports.

0.11.0
======

* Add Graphite.Report module to generate reports for easy printing

0.10.0
======

* Add interface to make irrelevant contents private

0.9.0
=====

* Initial extraction into a library
* Convert from OUnit2 to alcotest


test/test.ml
open Core_kernel
module Percentile = Graphite_async.Percentile

(* Should have tests for the fractal part *)

let assert_stats ?(percentiles = []) ~now ?total_obs ?avg ?sum ?period_obs t =
  let value ~name stats = List.Assoc.find ~equal:String.equal stats name in
  let stats = Percentile.get_stats t ~now in
  let percentile_key percentile =
    let p = Printf.sprintf "%.2f" percentile |> String.tr ~target:'.' ~replacement:'_' in
    Printf.sprintf "percentiles.60s.%s" p
  in
  let assert_value ~name expected =
    Alcotest.(check (option int)) name expected (value ~name stats)
  in
  assert_value ~name:"observations" total_obs;
  assert_value ~name:"percentiles.60s.avg" avg;
  assert_value ~name:"sum" sum;
  assert_value ~name:"percentiles.60s.observations" period_obs;
  List.iter percentiles ~f:(fun (percentile, value) ->
      assert_value ~name:(percentile_key percentile) value );
  Alcotest.(check int)
    "Too many percentiles returned"
    (List.length percentiles)
    (List.length stats - 4)

let test_empty_fractile () =
  let t = Percentile.init ~period:60 ~percentiles:[50.; 100.] in
  assert_stats
    t
    ~now:0
    ~total_obs:0
    ~avg:0
    ~sum:0
    ~period_obs:0
    ~percentiles:[50., Some 0; 100., Some 0]

let test_fractiles () =
  let t = Percentile.init ~period:60 ~percentiles:[50.; 100.] in
  Percentile.add_observation t ~now:10 1;
  Percentile.add_observation t ~now:20 2;
  Percentile.add_observation t ~now:30 3;
  Percentile.add_observation t ~now:40 4;
  Percentile.add_observation t ~now:50 5;
  assert_stats
    t
    ~now:60
    ~total_obs:5
    ~avg:3
    ~sum:15
    ~period_obs:5
    ~percentiles:[50., Some 3; 100., Some 5];
  assert_stats
    t
    ~now:90
    ~total_obs:5
    ~avg:4
    ~sum:15
    ~period_obs:3
    ~percentiles:[50., Some 4; 100., Some 5];
  assert_stats
    t
    ~now:120
    ~total_obs:5
    ~avg:0
    ~sum:15
    ~period_obs:0
    ~percentiles:[50., Some 0; 100., Some 0];
  (* Test adding old observations only adds to the summary *)
  Percentile.add_observation t ~now:30 3;
  Percentile.add_observation t ~now:40 4;
  Percentile.add_observation t ~now:50 5;
  assert_stats
    t
    ~now:120
    ~total_obs:8
    ~avg:0
    ~sum:27
    ~period_obs:0
    ~percentiles:[50., Some 0; 100., Some 0];
  Percentile.add_observation t ~now:100 3;
  Percentile.add_observation t ~now:100 4;
  Percentile.add_observation t ~now:100 5;
  let percentiles = [50., Some 4; 100., Some 5] in
  assert_stats t ~now:120 ~total_obs:11 ~avg:4 ~sum:39 ~period_obs:3 ~percentiles;
  (* Calling again with the same timestamp should yeild the same stats *)
  assert_stats t ~now:120 ~total_obs:11 ~avg:4 ~sum:39 ~period_obs:3 ~percentiles

let test_set =
  [ Alcotest.test_case "test fractile with no observations" `Quick test_empty_fractile;
    Alcotest.test_case "test fractiles" `Quick test_fractiles ]

let () = Alcotest.run Caml.__MODULE__ ["test_set", test_set]


lib/graphite_async.ml
open Core_kernel
open Async
module Amqp = Amqp_client_async

let now () = Unix.gettimeofday () |> Int.of_float

let timer_ms () =
  let start_ts = Time.now () in
  fun () ->
    let end_ts = Time.now () in
    Time.diff end_ts start_ts

module Percentile = struct
  type ts = int

  type value = int

  type t = {
    period : int;
    percentiles : float list;
    mutable total_observations : int;
    mutable sum : value;
    mutable observations : (value * ts) list (* Ordered by timestamp *)
  }

  let init ~period ~percentiles =
    {period; percentiles; total_observations = 0; sum = 0; observations = []}

  let expire ~now t =
    t.observations
    <- List.take_while ~f:(fun (_, ts) -> ts + t.period >= now) t.observations

  let average observations =
    match Array.length observations with
    | 0 -> 0
    | n -> Array.fold ~init:0 ~f:( + ) observations / n

  let make_percentile_key period value = Printf.sprintf "percentiles.%ds.%s" period value

  let make_percentile_key_float period percentile =
    Printf.sprintf "%.2f" percentile
    |> String.tr ~target:'.' ~replacement:'_'
    |> make_percentile_key period

  let get_stats ~now t =
    expire ~now t;
    let sorted_observations =
      let observations = Array.of_list_map ~f:fst t.observations in
      Array.sort ~compare:Int.compare observations;
      observations
    in
    let observation_count = Array.length sorted_observations in
    let get_percentile_value percentile =
      match observation_count with
      | 0 -> 0 (* Default to zero if there no observations *)
      | n ->
          let ceil_index =
            percentile *. Float.of_int n /. 100.0 |> Float.round_up |> Int.of_float
          in
          sorted_observations.(ceil_index - 1)
    in
    let percentile_values =
      List.map t.percentiles ~f:(fun percentile ->
          make_percentile_key_float t.period percentile, get_percentile_value percentile
      )
    in
    ("sum", t.sum)
    :: ("observations", t.total_observations)
    :: (make_percentile_key t.period "observations", observation_count)
    :: (make_percentile_key t.period "avg", average sorted_observations)
    :: percentile_values

  let add_observation t ~now value =
    t.observations <- (value, now) :: t.observations;
    t.total_observations <- t.total_observations + 1;
    t.sum <- t.sum + value
end

type connection =
  | Mock
  | Real of {
    channel : Amqp.Channel.no_confirm Amqp.Channel.t;
    exchange : [`Topic of string] Amqp.Exchange.t;
    prefix : string; }

type t = {
  connection : connection;

  metrics : (string, int) Hashtbl.t;
  percentile_period : int;
  (* At some point we should allow multiple periods *)
  percentiles : float list;
  percentile_data : (string, Percentile.t) Hashtbl.t;
  init_time : Time.t
}

type stat = {
  path : string;
  value : int;
  ts : int
}

let send_raw t stats =
  match t with
  | { connection=Mock; _ } -> return ()
  | { connection=Real t; _ } ->
    let msg =
      stats
      |> List.map ~f:(fun stat ->
            Printf.sprintf "%s.%s %d %d" t.prefix stat.path stat.value stat.ts )
      |> String.concat ~sep:"\n"
    in
    Amqp.Exchange.publish
      t.channel
      t.exchange
      ~routing_key:"graphite"
      (Amqp.Message.make msg)
    >>= fun `Ok -> return ()

let send t () =
  let now = now () in
  Hashtbl.fold t.percentile_data ~init:[] ~f:(fun ~key ~data:fractile acc ->
      Percentile.get_stats ~now fractile
      |> List.fold ~init:acc ~f:(fun acc (k, v) ->
             {path = Printf.sprintf "%s.%s" key k; value = v; ts = now} :: acc ) )
  |> (fun acc ->
       Hashtbl.fold t.metrics ~init:acc ~f:(fun ~key ~data acc ->
           {path = key; value = data; ts = now} :: acc ) )
  |> send_raw t

let flush t = send t ()

module Report = struct
  type metric = {
    observations : int;
    obs_rate : float;
    rate : float option;
    avg : float option;
    sum : int option
  }
  [@@deriving show]

  type t = (string * metric) list [@@deriving show]
end

let report t =
  let time_since_init = Time.diff (Time.now ()) t.init_time in
  let percentile_data =
    t.percentile_data
    |> Hashtbl.to_alist
    |> List.map ~f:(fun (key, fractile) ->
           ( key,
             Report.
               { observations = fractile.Percentile.total_observations;
                 obs_rate =
                   Float.(
                     of_int fractile.Percentile.total_observations
                     / Time.Span.to_sec time_since_init);
                 rate =
                   Some
                     Float.(
                       of_int fractile.Percentile.sum / Time.Span.to_sec time_since_init);
                 avg =
                   Some
                     Float.(
                       of_int fractile.Percentile.sum
                       / of_int fractile.Percentile.total_observations);
                 sum = Some fractile.Percentile.sum } ) )
  in
  let metric_data =
    t.metrics
    |> Hashtbl.to_alist
    |> List.map ~f:(fun (key, metric) ->
           ( key,
             Report.
               { observations = metric;
                 obs_rate = Float.of_int metric /. Time.Span.to_sec time_since_init;
                 rate = None;
                 avg = None;
                 sum = None } ) )
  in
  percentile_data @ metric_data

let init_mock ?(percentile_period = 60) ?(percentiles = [50.; 90.; 95.; 99.]) () =
  { connection=Mock;
    metrics = Hashtbl.Poly.create ();
    percentile_period;
    percentiles;
    percentile_data = Hashtbl.Poly.create ();
    init_time = Time.now () }

let init ?(updates_per_minute = 2) ?(percentile_period = 60)
    ?(percentiles = [50.; 90.; 95.; 99.]) ~prefix connection
  =
  let%bind channel =
    Amqp.Connection.open_channel ~id:"graphite" Amqp.Channel.no_confirm connection
  in
  let%bind exchange =
    Amqp.Exchange.declare ~durable:true channel Amqp.Exchange.topic_t "metrics"
  in
  let connection = Real {
    channel;
    exchange;
    prefix;
  }
in
  let t =
    { connection;
      metrics = Hashtbl.Poly.create ();
      percentile_period;
      percentiles;
      percentile_data = Hashtbl.Poly.create ();
      init_time = Time.now () }
  in
  (* Start the sender *)
  let send_frequency = Time.Span.of_int_sec (60 / updates_per_minute) in
  Clock.every' send_frequency (send t);
  return t

let incr ?by ~key t = Hashtbl.incr ?by t.metrics key

let decr t ~key value = Hashtbl.decr ~by:value t.metrics key

let set t ~key value = Hashtbl.set t.metrics ~key ~data:value

let remove t ~key = Hashtbl.remove t.metrics key

let add_percentile_observation t ~key value =
  Hashtbl.find_or_add t.percentile_data key ~default:(fun () ->
      Percentile.init ~period:t.percentile_period ~percentiles:t.percentiles )
  |> fun f -> Percentile.add_observation f ~now:(now ()) value

module Result = struct
  let time ~graphite ?callback ~key ~f v =
    let timer = timer_ms () in
    let%bind r = f v in
    let time_span = timer () in
    Core.Option.iter callback ~f:(fun callback -> callback time_span v r);
    let sub_key =
      match r with
      | Ok _ -> "ok"
      | Error _ -> "error"
    in
    add_percentile_observation
      graphite
      ~key:(Printf.sprintf "%s.%s" key sub_key)
      (time_span |> Time.Span.to_ms |> int_of_float);
    return r
end

module Option = struct
  let time ~graphite ?callback ~key ~f v =
    let timer = timer_ms () in
    let%bind r = f v in
    let time_span = timer () in
    Core.Option.iter callback ~f:(fun callback -> callback time_span v r);
    let sub_key =
      match r with
      | Some _ -> "ok"
      | None -> "error"
    in
    add_percentile_observation
      graphite
      ~key:(Printf.sprintf "%s.%s" key sub_key)
      (time_span |> Time.Span.to_ms |> int_of_float);
    return r
end

module Deferred = struct
  let time ~graphite ?callback ~key ~f v =
    let timer = timer_ms () in
    let%bind r = f v in
    let time_span = timer () in
    Core.Option.iter callback ~f:(fun callback -> callback time_span v r);
    add_percentile_observation
      graphite
      ~key:(Printf.sprintf "%s.ok" key)
      (time_span |> Time.Span.to_ms |> int_of_float);
    return r

  let keyed_time ~graphite ?callback ~key ~f v =
    let timer = timer_ms () in
    let%bind chunk, r = f v in
    let time_span = timer () in
    Core.Option.iter callback ~f:(fun callback -> callback time_span v (chunk, r));
    add_percentile_observation
      graphite
      ~key:(Printf.sprintf "%s.%s" key chunk)
      (time_span |> Time.Span.to_ms |> int_of_float);
    return r
end
